### PR TITLE
mvcc: restore into tree index with one key index

### PR DESCRIPTION
**Pull request by heyitsanthony**
Clobbering the mvcc kvindex with new keyIndexes for each restore
chunk would cause index corruption by dropping historical information.

### CODE REVIEW

**Comment by xiang90** [comment]
will this fix https://github.com/coreos/etcd/issues/8048?

**Comment by heyitsanthony** [comment]
@xiang90 yes, appears to be fixed; could repro on master, can't repro with the patch

File: mvcc/kvstore.go [hunk 1]
```diff
@@ -334,38 +326,73 @@ func (s *store) restore() error {
...
+func restoreIntoIndex(idx index) (chan<- revKeyValue, <-chan int64) {
+	rkvc, revc := make(chan revKeyValue, restoreChunkKeys), make(chan int64, 1)
+	go func() {
+		currentRev := int64(1)
+		defer func() { revc <- currentRev }()
+		// restore the tree index from streaming the unordered index.
+		kiCache := make(map[string]*keyIndex, restoreChunkKeys)
+		for rkv := range rkvc {
+			ki, ok := kiCache[rkv.kstr]
+			// purge cache if too many keys and missing
```
Comment (gyuho) [inline mvcc/kvstore.go]: `s/many keys and/many keys are/`?

Comment (heyitsanthony) [inline mvcc/kvstore.go]: missing as in rkv.kstr is missing, but can clarify

**Pull request merged**